exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/MergingReceiverCreator.java

```diff
@@ -44,6 +44,11 @@ public MergingRecordBatch getBatch(ExecutorFragmentContext context,
     assert bufHolder != null : "IncomingBuffers must be defined for any place a receiver is declared.";
     RawBatchBuffer[] buffers = bufHolder.getBuffers(receiver.getOppositeMajorFragmentId());
 
-    return new MergingRecordBatch(context, receiver, buffers);
+    MergingRecordBatch mergeReceiver = new MergingRecordBatch(context, receiver, buffers);
+
+    // Register this operator's buffer allocator so that incoming buffers are owned by this allocator
+    bufHolder.setOprAllocator(receiver.getOppositeMajorFragmentId(), mergeReceiver.getOprAllocator());
```

Comment:
Consider moving registration of the buffer allocator inside `MerginRecordBatch` constructor (change constructor to accept `ExchangeFragmentContext` and `MergingReceiverPOP` only).

Reply:
sure.